miner/payload_building.go

```diff
@@ -0,0 +1,160 @@
+	FeeRecipient common.Address // The provided recipient address for collecting transaction fee
+	Random       common.Hash    // The provided randomness value
+}
+
+// Payload wraps the built payload(block waiting for sealing). According to the
+// engine-api specification, EL should build the initial version of the payload
+// which has an empty transaction set and then keep update it in order to maximize
+// the revenue. Therefore, the empty-block here is always available and full-block
+// will be set/updated afterwards.
+type Payload struct {
+	empty *types.Block
+	full  *types.Block
+	stop  chan struct{}
+	lock  *sync.RWMutex
+	cond  *sync.Cond
+}
+
+// newPayload initializes the payload object.
+func newPayload() *Payload {
+	lock := new(sync.RWMutex)
+	return &Payload{
+		stop: make(chan struct{}),
+		lock: lock,
+		cond: sync.NewCond(lock),
+	}
+}
+
+// update updates the full-block with latest built version.
+func (payload *Payload) update(block *types.Block) {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	select {
+	case <-payload.stop:
+		return // reject stale update
+	default:
+	}
+	payload.full = block
+	payload.cond.Broadcast() // fire signal for notifying full block
+}
+
+// Resolve returns the latest built payload and also terminates the background
+// thread for updating payload. It's safe to be called multiple times.
+func (payload *Payload) Resolve() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	select {
+	case <-payload.stop:
+	default:
+		close(payload.stop)
+	}
+	if payload.full != nil {
+		return beacon.BlockToExecutableData(payload.full)
+	}
+	return beacon.BlockToExecutableData(payload.empty)
+}
+
+// ResolveEmpty is basically identical to Resolve, but it expects empty block only.
+// It's only used in tests.
+func (payload *Payload) ResolveEmpty() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	return beacon.BlockToExecutableData(payload.empty)
+}
+
+// ResolveFull is basically identical to Resolve, but it expects full block only.
+// It's only used in tests.
+func (payload *Payload) ResolveFull() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	if payload.full == nil {
+		select {
+		case <-payload.stop:
+			return nil
+		default:
+		}
+		payload.cond.Wait()
+	}
+	return beacon.BlockToExecutableData(payload.full)
+}
+
+// buildPayload builds the payload according to the provided parameters.
+func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
+	// Build the initial version with no transaction included. It should be fast
+	// enough to run. The empty payload can at least make sure there is something
+	// to deliver for not missing slot.
+	empty, err := w.getSealingBlock(args.Parent, args.Timestamp, args.FeeRecipient, args.Random, true)
+	if err != nil {
+		return nil, err
+	}
+	// Construct a payload object for return.
+	payload := newPayload()
```

Comment:
Maybe change it like thisL 
```
payload := newPayload(empty)
```
and make `newPayload(*types.Block)` . Makes it more obvious that this is not a race-condition, and generally nicer not to modify internals like this.

Reply:
fixed.